src/readblock.c

```diff
@@ -0,0 +1,299 @@
+    while (0!=strncmp (tag, F->next_args, n))
+        if (0==nextline (F))
+            return 0;
+    return 1;
+}
+
+
+/***************************************************************************************/
+static int step_into_gie_block (ffio *F) {
+/****************************************************************************************
+    Make sure we're inside a <gie>-block
+****************************************************************************************/
+    /* Already inside */
+    if (F->level % 2)
+        return 1;
+
+    if (0==locate_tag (F, "<gie>"))
+        return 0;
+
+    while (0!=strncmp ("<gie>", F->next_args, 5)) {
+        printf ("skipping [%s]\n", F->next_args);
+        F->next_args[0] = 0;
+        if (feof (F->f))
+            return 0;
+        if (0==fgets (F->next_args, (int) F->next_args_size - 1, F->f))
+            return 0;
+        pj_chomp (F->next_args);
+        F->next_lineno++;
+    }
+    F->level++;
+
+    /* We're ready at the start - now step into the block */
+    return nextline (F);
+}
+
+
+/***************************************************************************************/
+static int skip_to_next_tag (ffio *F) {
+/****************************************************************************************
+    Skip forward to the next command tag
+****************************************************************************************/
+    const char *c;
+    if (0==step_into_gie_block (F))
+        return 0;
+
+    c = at_tag (F);
+
+    /* If not already there - get there */
+    while (!c) {
+        if (0==nextline (F))
+            return 0;
+        c = at_tag (F);
+    }
+
+    /* If we reached the end of a <gie> block, locate the next and retry */
+    if (0==strcmp (c, "</gie>")) {
+        F->level++;
+        if (feof (F->f))
+            return 0;
+        if (0==step_into_gie_block (F))
+            return 0;
+        F->args[0] = 0;
+        return skip_to_next_tag (F);
+    }
+    F->lineno = F->next_lineno;
+
+    return 1;
+}
+
+
+static int append_args (ffio *F) {
+    size_t skip_chars = 0;
+    size_t next_len = strlen (F->next_args);
+    size_t args_len = strlen (F->args);
+    const char *tag = at_tag (F);
+
+    if (tag)
+        skip_chars = strlen (tag);
+
+    if (F->args_size < args_len + next_len - skip_chars + 1) {
+        void *p = realloc (F->args, 2 * F->args_size);
+        if (0==p)
+            return 0;
+        F->args = p;
+        F->args_size = 2 * F->args_size;
+    }
+
+    F->args[args_len] = ' ';
+    strcpy (F->args + args_len + 1,  F->next_args + skip_chars);
+
+    F->next_args[0] = 0;
+    return 1;
+}
+
+
+
+
+
+/***************************************************************************************/
+static int get_inp (ffio *F) {
+/****************************************************************************************
+    The primary command reader for gie. May be useful in the init-file reader as well.
```

Comment:
Sorry, what does this do?  What state changes to F does this cause?  Why would it be use in the init-file reader?

Reply:
Improved and increased comment to address this